org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/pom/ReImportMavenProjectsHandler.java

```diff
@@ -0,0 +1,175 @@
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.jobs.IJobChangeEvent;
+import org.eclipse.core.runtime.jobs.IJobManager;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.jobs.JobChangeAdapter;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.managers.ProjectsManager;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
+
+/** @author Mykola Morhun */
+public class ReImportMavenProjectsHandler {
+  public static long REIMPORT_TIMEOUT = 60_000L;
+
+  private static final Gson gson =
+      new GsonBuilder()
+          .registerTypeAdapterFactory(new CollectionTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EitherTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
+          .create();
+
+  private static final PreferenceManager preferenceManager = new PreferenceManager();
+  private static final ProjectsManager projectsManager = new ProjectsManager(preferenceManager);
+  private static final IJobManager jobManager = Job.getJobManager();
+  private static final JobChangedListener jobChangedListener = new JobChangedListener();
+  private static final Lock lock = new ReentrantLock();
+
+  private static CountDownLatch jobsToBeFinished;
+  // name -> path
+  private static Map<String, String> projectsToBeUpdated = new ConcurrentHashMap<>();
+  // set of paths to updates projects; path -> ""
+  private static Map<String, String> updatedProjectsPaths = new ConcurrentHashMap<>();
+
+  /**
+   * Updates given maven projects.
+   *
+   * @param arguments contains ReImportMavenProjectsCommandParameters in the first element
+   * @param progressMonitor progress monitor
+   * @return URIs of updated projects
+   */
+  public static List<String> reImportMavenProjects(
+      List<Object> arguments, IProgressMonitor progressMonitor) {
+
+    if (lock.tryLock()) {
+      try {
+        updatedProjectsPaths.clear();
+
+        ReImportMavenProjectsCommandParameters parameters =
+            gson.fromJson(
+                gson.toJson(arguments.get(0)), ReImportMavenProjectsCommandParameters.class);
+
+        ensureNotCancelled(progressMonitor);
+        return new ArrayList<>(
+            doReImportMavenProjects(parameters.getProjectsToUpdate(), progressMonitor));
+      } catch (InterruptedException e) {
+        throw new OperationCanceledException(e.getMessage());
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    throw new RuntimeException("Reimport is already in progress.");
```

Comment:
I am not sure throwing exception is the right choice. Let's return nothing.

Reply:
Then user could be confused. Imagine, that module A and B are under reimport. User requests reimport of C and D and gets nothing into result. No success nor error. Weird.

Reply:
Add in the queue?

Reply:
Definitely exception will confuse user more)

Reply:
Well, exception has appropriate message and user will know what happened and why.
I am not sure that queue is a good idea. Especially when no progress bar visible.

Reply:
You will definitely surprise user with an exception :)

Reply:
@tsmaeder @svor we need your opinion

Reply:
i agree with @tolusha, the queue is better. It isn't an exception and i think we'll have a progressor in the feature

Reply:
The problem with the queue is that we'll time out on the Che side. The _real_ solution is to have async notifications for long running operations. What I think would be the worst solution is to add to the queue and return early from this method: the user would think the operation is done, when it is not.
For mutual exclusion, think about using the locking mechanism for jobs that Eclipse already provides (see Job#getSchedulingRule())